src/Xamarin.Android.Build.Tasks/Tasks/AsyncTask.cs

```diff
@@ -214,6 +217,24 @@ public void LogWarning (string message)
 			}
 		}
 
+		public void LogCustomBuildEvent (CustomBuildEventArgs e)
+		{
+			if (UIThreadId == Thread.CurrentThread.ManagedThreadId) {
+#pragma warning disable 618
+				BuildEngine.LogCustomEvent (e);
+				return;
+#pragma warning restore 618
+			}
+
+			lock (customMessageQueue.SyncRoot) {
```

Comment:
This feels like an opportunity for cleanup/a new helper method, as `LogCustomBuildEvent()` looks very similar to `LogMessage()` looks very similar to `LogError()` looks...

Particularly around the locking and message queuing.

Reply:
I can rework this into a helper. Each queue still needs its own reset event though.